keepalive.go

```diff
@@ -19,7 +22,42 @@ type KeepAlive struct {
 	// WaitForPong how long it will waits for pong response.
 	WaitForPong time.Duration
 	// NewRetryPolicy creates retry policy for the connection when ping fails.
-	NewRetryPolicy func() BackOff
+	NewRetryPolicy RetryFuncFactory
+}
+
+const minDuration = time.Millisecond * 50
+
+// MakeKeepAlive creates a policy that detects dropped connections within the connTimeout limit (3 pings in a row failed)
+func MakeKeepAlive(connTimeout time.Duration) (KeepAlive, error) {
+	duration := connTimeout / 6
+	if duration < minDuration {
+		return KeepAlive{}, fmt.Errorf("connTimeout %v it too small: must be greater than %v", connTimeout, minDuration*6)
+	}
+	return KeepAlive{
+		Enable:      true,
+		Interval:    duration,
+		WaitForPong: duration,
+		NewRetryPolicy: func() RetryFunc {
+			now := time.Now()
+			// try 2 times to send ping after fails (wait, pong, wait, pong)
+			return func() (time.Time, error) {
+				c := time.Now()
+				if c.Before(now.Add(duration * 4)) {
+					return c.Add(duration), nil
+				}
+				return time.Time{}, ErrKeepAliveDeadlineExceeded
+			}
```

Comment:
To prevent drifting:

```go
start := time.Now()
attempt := 0
return func() (time.Time, error) {
	attempt++
	// Try to send ping and wait for pong 2 more times
	if time.Since(start) <= 2 * 2 * duration {
		return start.Add(attempt * 2 * duration), nil
	}
	return time.Time{}, ErrKeepAliveDeadlineExceeded
}
```